### PR TITLE
More E2E tests and general cleanup

### PR DESCRIPTION
**Pull request by Bslabe123**
#### What this PR does / why we need it
Added E2E tests around some common use cases

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -37,7 +42,7 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
 				GenerateName: "test-ns-",
 			},
 		}
-		Expect(k8sClient.Create(ctx, ns)).To(Succeed())
+		Eventually(k8sClient.Create(ctx, ns)).Should(Succeed())
```
Comment (kerthcet) [inline test/e2e/e2e_test.go:45]: Use Eventually is ok, do you meet failure ever especially in Bofore?

File: test/e2e/e2e_test.go [hunk 2]
```diff
@@ -51,23 +56,169 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
 	})
 
 	It("Can deploy lws", func() {
-		lws := testing.BuildLeaderWorkerSet(ns.Name).Obj()
-		Expect(k8sClient.Create(ctx, lws)).To(Succeed())
-		testing.ExpectLeaderWorkerSetAvailable(ctx, k8sClient, lws, "All replicas are ready")
```
Comment (kerthcet) [inline test/e2e/e2e_test.go:56]: Why remove this? How do you know that lws is ready?

File: test/testutils/util.go [hunk 3]
```diff
@@ -35,8 +35,18 @@ import (
 	acceleratorutils "sigs.k8s.io/lws/pkg/utils/accelerators"
 )
 
+func CreateLws(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet) {
+	Eventually(k8sClient.Create(ctx, lws)).Should(Succeed())
+	Eventually(func() error {
```
Comment (kerthcet) [inline test/testutils/util.go:40]: Maybe we should expect lws is available here? If we create successfully, I can't image why we'll not get the object here unless the apiserver/etcd is down.

File: test/e2e/e2e_test.go [hunk 4]
```diff
@@ -51,23 +56,169 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
+		leaderPodSpec := testing.MakeLeaderPodSpecWithTPUResource()
+		workerPodSpec := testutils.MakeWorkerPodSpecWithTPUResource()
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(2).Size(2).LeaderTemplateSpec(leaderPodSpec).WorkerTemplateSpec(workerPodSpec).Obj()
+		testing.CreateLws(ctx, k8sClient, lws)
+
+		lwsPods := &corev1.PodList{}
+		testing.GetPods(ctx, lws, k8sClient, lwsPods)
+
+		Eventually(func() (bool, error) {
+			var allPods corev1.PodList
```
Comment (kerthcet) [inline test/e2e/e2e_test.go:97]: allPods here for what?

File: test/e2e/e2e_test.go [hunk 5]
```diff
@@ -51,23 +56,169 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
 
 	It("Can deploy lws", func() {
-		lws := testing.BuildLeaderWorkerSet(ns.Name).Obj()
-		Expect(k8sClient.Create(ctx, lws)).To(Succeed())
-		testing.ExpectLeaderWorkerSetAvailable(ctx, k8sClient, lws, "All replicas are ready")
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(3).Obj()
+		testing.CreateLws(ctx, k8sClient, lws)
+	})
+
+	It("Can deploy lws with 'replicas', 'size', and 'restart policy' set", func() {
```
Comment (kerthcet) [inline test/e2e/e2e_test.go:63]: Maybe we can combine with the upper case because e2e test is heavy.

File: test/e2e/e2e_test.go [hunk 6]
```diff
@@ -51,23 +56,169 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
+
+		for _, p := range lwsPods.Items {
+			Expect(testing.HasTPUEnvVarsPopulated(p)).To(BeFalse())
+		}
+	})
+
+	It("Pod restart will not recreate the pod group when restart policy is Default", func() {
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(1).Size(3).RestartPolicy(v1.DefaultRestartPolicy).Obj()
+		testing.CreateLws(ctx, k8sClient, lws)
+		testing.GetPods(ctx, lws, k8sClient, &corev1.PodList{})
```
Comment (kerthcet) [inline test/e2e/e2e_test.go:128]: We do have `ExpectValidPods`.

File: test/e2e/e2e_test.go [hunk 7]
```diff
@@ -51,23 +56,169 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
+
+		var leaderPod corev1.Pod
```
Comment (kerthcet) [inline test/e2e/e2e_test.go:130]: I guess this is too detailed? Seems we're testing the statefulset logic here. Better to relate to `RestartPolicy`.

**Review by kerthcet — COMMENTED** [review]
Didn't review the whole logic of the PR, just some comments at the first glance.

File: test/e2e/e2e_test.go [hunk 8]
```diff
@@ -51,23 +56,174 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
+		Eventually(k8sClient.List(ctx, &workers, client.InNamespace(lws.Namespace))).Should(Succeed())
+
+		// Get all lws pod UIDs now and compare them to the UIDs after deletion
+		// With RecreateGroupOnPodRestart they should all be different
+		initialPodUIDs := make(map[types.UID]struct{})
+		for _, w := range workers.Items {
+			initialPodUIDs[w.UID] = struct{}{}
+		}
+
+		Eventually(k8sClient.Delete(ctx, &firstWorker)).Should(Succeed())
```
Comment (ahg-g) [inline test/e2e/e2e_test.go:203]:
```
k8sClient.Delete(context.Background(), &corev1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: lws.Namespace,
			Name:      lws.Name + "-0-1",
		},
	})
```

File: test/e2e/e2e_test.go [hunk 9]
```diff
@@ -51,23 +56,174 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
+
+	It("Pod restart will delete the pod group when restart policy is RecreateGroupOnPodRestart", func() {
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(1).Size(3).RestartPolicy(v1.RecreateGroupOnPodRestart).Obj()
+		testing.MustCreateLws(ctx, k8sClient, lws)
+		testing.ExpectLeaderWorkerSetAvailable(ctx, k8sClient, lws, "All replicas are ready")
+
+		testing.ExpectValidPods(ctx, k8sClient, lws, &corev1.PodList{})
+
+		var leaderPod corev1.Pod
+		Eventually(k8sClient.Get(ctx, types.NamespacedName{Name: lws.Name + "-0", Namespace: lws.Namespace}, &leaderPod)).Should(Succeed())
```
Comment (ahg-g) [inline test/e2e/e2e_test.go:190]: we don't need to get this pod? the name is known.

File: test/e2e/e2e_test.go [hunk 10]
```diff
@@ -51,23 +56,174 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
+		var firstWorker corev1.Pod
+		Eventually(k8sClient.Get(ctx, types.NamespacedName{Name: leaderPod.Name + "-1", Namespace: leaderPod.Namespace}, &firstWorker)).Should(Succeed())
```
Comment (ahg-g) [inline test/e2e/e2e_test.go:192]: we don't need to get the pod to delete it, see the comment below

File: test/e2e/e2e_test.go [hunk 11]
```diff
@@ -51,23 +56,174 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
+		Eventually(k8sClient.Delete(ctx, &firstWorker)).Should(Succeed())
+
+		Eventually(func() (int, error) {
+			var leaders corev1.PodList
+			err := k8sClient.List(ctx, &leaders, client.InNamespace(lws.Namespace), &client.MatchingLabels{leaderworkerset.WorkerIndexLabelKey: "0"})
+			if err != nil {
+				return -1, err
+			}
+			return len(leaders.Items), nil
+		}, timeout, interval).Should(Equal(int(*lws.Spec.Replicas)))
```
Comment (ahg-g) [inline test/e2e/e2e_test.go:212]: what are we checking here? I don't think you need this.

File: test/e2e/e2e_test.go [hunk 12]
```diff
@@ -51,23 +56,174 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
...
+
+		Eventually(func() (int, error) {
+			err := k8sClient.List(ctx, &workers, client.InNamespace(lws.Namespace))
+			if err != nil {
+				return -1, err
+			}
```
Comment (ahg-g) [inline test/e2e/e2e_test.go:218]: check that the list length is equal to the expected number of pods

File: test/e2e/e2e_test.go (hunk 7 above)
Comment (ahg-g) [reply test/e2e/e2e_test.go:133]: +1; I don't think we want this test.

File: test/e2e/e2e_test.go (hunk 1 above)
Comment (liurupeng) [reply test/e2e/e2e_test.go:45]: +1, probably remove this change

**Comment by ahg-g** [comment]
/lgtm

leaving the approve to either @kerthcet or @liurupeng

File: test/e2e/e2e_test.go [hunk 13]
```diff
@@ -16,58 +16,137 @@ limitations under the License.
...
 
-	It("Can deploy lws", func() {
-		lws := testing.BuildLeaderWorkerSet(ns.Name).Obj()
-		Expect(k8sClient.Create(ctx, lws)).To(Succeed())
+	ginkgo.It("Can deploy lws with 'replicas', 'size', and 'restart policy' set", func() {
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(4).Size(5).RestartPolicy(v1.RecreateGroupOnPodRestart).Obj()
+		testing.MustCreateLws(ctx, k8sClient, lws)
 		testing.ExpectLeaderWorkerSetAvailable(ctx, k8sClient, lws, "All replicas are ready")
+		testing.GetStatefulSets(ctx, lws, k8sClient, &appsv1.StatefulSetList{})
+
```
Comment (liurupeng) [inline test/e2e/e2e_test.go:64]: we could check the expected annotation/labels are added as well

**Pull request merged**